File: Part7/C/LoopConcepts/CompilerErrors.c
#include <limits.h>
#include <stdio.h>
#include <string.h>
#include <stdbool.h>

int main()
{
    // ERROR: Missing parentheses around for loop parameters
    printf("Missing parentheses in for loop\n");
    for int i = 0; i < 10; i++
    {
        printf("%d\n", i);
    }

    // ERROR: Using assignment (=) instead of comparison (==) in condition
    int counter = 0;
    while (counter = 10)
    {
        printf("%d\n", counter);
        counter++;
    }

    // ERROR: Missing braces and semicolon creates wrong association
    for (int i = 0; i < 5; i++)
        printf("Loop iteration\n");
        printf("This should be in loop\n");

    // ERROR: Using single = instead of == for comparison
    int value = 5;
    while (value = 0)
    {
        printf("This won't print\n");
        value--;
    }

    // ERROR: Missing increment in for loop
    for (int i = 0; i < 10)
    {
        printf("%d\n", i);
    }

    // ERROR: Missing condition in for loop
    for (int i = 0; ; i++)
    {
        printf("%d\n", i);
    }

    // ERROR: Using wrong increment operator (-- instead of ++)
    for (int i = 0; i < 10; i--)
    {
        printf("%d\n", i);
    }

    // ERROR: Missing condition expression entirely
    int count = 0;
    while ()
    {
        count++;
        if (count > 10) break;
    }

    // ERROR: Using 'do' without 'while' at the end
    int num = 0;
    do
    {
        printf("%d\n", num);
        num++;
    }

    // ERROR: Extra semicolon after while in do-while
    int x = 0;
    do
    {
        printf("%d\n", x);
        x++;
    }
    while (x < 5);
    ;

    // ERROR: Missing semicolon after do-while
    int y = 0;
    do
    {
        printf("%d\n", y);
        y++;
    }
    while (y < 5)

    // ERROR: Using curly braces in for loop parameter section
    for {int i = 0; i < 10; i++}
    {
        printf("%d\n", i);
    }

    // ERROR: Undeclared variable (missing type declaration)
    for (i = 0; i < 10; i++)
    {
        printf("%d\n", i);
    }

    // ERROR: Missing opening brace
    for (int i = 0; i < 10; i++)
        printf("%d\n", i);
    }

    // ERROR: Missing closing brace
    for (int i = 0; i < 10; i++)
    {
        printf("%d\n", i);

    // ERROR: Using := instead of = for assignment
    int idx := 0;
    while (idx < 10)
    {
        printf("%d\n", idx);
        idx++;
    }

    // ERROR: Missing parentheses around while condition
    int k = 0;
    while k < 10
    {
        printf("%d\n", k);
        k++;
    }

    // ERROR: Using 'then' keyword (from other languages)
    int m = 0;
    while (m < 10) then
    {
        printf("%d\n", m);
        m++;
    }

    // ERROR: Using ++ twice in increment
    for (int i = 0; i < 10; ++i++)
    {
        printf("%d\n", i);
    }

    // ERROR: Missing opening parenthesis in while
    int n = 0;
    while n < 10)
    {
        printf("%d\n", n);
        n++;
    }

    // ERROR: Missing closing parenthesis in while
    int p = 0;
    while (p < 10
    {
        printf("%d\n", p);
        p++;
    }

    // ERROR: Using 'end' or 'end for' (from other languages)
    for (int i = 0; i < 10; i++)
    {
        printf("%d\n", i);
    }
    end for;

    // ERROR: Trying to use range-based for loop with 'in' keyword
    int arr[] = {1, 2, 3, 4, 5};
    for (int element in arr)
    {
        printf("%d\n", element);
    }

    // ERROR: Using 'foreach' keyword (not valid in C)
    foreach (int element in arr)
    {
        printf("%d\n", element);
    }

    // ERROR: Multiple semicolons in for loop
    for (int i = 0;; i < 10;; i++)
    {
        printf("%d\n", i);
    }

    // ERROR: Using equals (==) instead of assignment (=) in initialization
    for (int i == 0; i < 10; i++)
    {
        printf("%d\n", i);
    }

    // ERROR: Missing semicolon after initialization
    for (int i = 0 i < 10; i++)
    {
        printf("%d\n", i);
    }

    // ERROR: Missing semicolon after condition
    for (int i = 0; i < 10 i++)
    {
        printf("%d\n", i);
    }

    // ERROR: Using 'continue' with label (not supported in C)
    for (int i = 0; i < 10; i++)
    {
        if (i % 2 == 0)
            continue outer;
        printf("%d\n", i);
    }

    // ERROR: Using 'break' with label (not supported in C)
    outer:
    for (int i = 0; i < 10; i++)
    {
        for (int j = 0; j < 10; j++)
        {
            if (j == 5)
                break outer;
        }
    }

    // ERROR: Declaring variable inside while condition
    while (int z = getchar())
    {
        printf("%c", z);
    }

    // ERROR: Using += with incorrect spacing
    for (int i = 0; i < 10; i+ =1)
    {
        printf("%d\n", i);
    }

    // ERROR: Trying to declare multiple different types in one for loop
    for (int i = 0, float j = 0.0; i < 10; i++, j++)
    {
        printf("%d %f\n", i, j);
    }

    // ERROR: Missing comma between multiple variable declarations
    for (int i = 0 j = 0; i < 10; i++, j++)
    {
        printf("%d %d\n", i, j);
    }

    // ERROR: Using 'loop' keyword (from Ada)
    loop (int i = 0; i < 10; i++)
    {
        printf("%d\n", i);
    }

    // ERROR: Using colon instead of semicolon
    for (int i = 0: i < 10: i++)
    {
        printf("%d\n", i);
    }

    // ERROR: Square brackets instead of parentheses
    for [int i = 0; i < 10; i++]
    {
        printf("%d\n", i);
    }

    // ERROR: Using float in loop comparison (implicit conversion warning/error)
    for (int i = 0; i < 10.5; i++)
    {
        printf("%d\n", i);
    }

    // ERROR: Using -> instead of ++ for increment
    for (int i = 0; i < 10; i->)
    {
        printf("%d\n", i);
    }

    // ERROR: Missing 'do' keyword in do-while
    int val = 0;
    {
        printf("%d\n", val);
        val++;
    }
    while (val < 5);

    // ERROR: Using 'repeat' and 'until' (from Pascal)
    int r = 0;
    repeat
    {
        printf("%d\n", r);
        r++;
    }
    until (r >= 10);

    // ERROR: Nested for without proper bracing causing declaration in wrong scope
    for (int i = 0; i < 3; i++)
    for (int j = 0; j < 3; j++)
    int product = i * j;
    printf("%d\n", product);

    // ERROR: Using double equals in increment
    for (int i = 0; i < 10; i == i + 1)
    {
        printf("%d\n", i);
    }

    // ERROR: Conflicting increment/decrement operators
    for (int i = 0; i < 10; ++i--)
    {
        printf("%d\n", i);
    }

    // ERROR: Using 'as' keyword for type (from other languages)
    for (i as int = 0; i < 10; i++)
    {
        printf("%d\n", i);
    }

    // ERROR: Using 'elsif' or 'elif' in loop (wrong context/language)
    for (int i = 0; i < 10; i++)
    elsif (i % 2 == 0)
    {
        printf("Even\n");
    }

    // ERROR: Missing while keyword, just condition
    int w = 0;
    (w < 10)
    {
        printf("%d\n", w);
        w++;
    }

    return 0;
}

File: Part7/Fortran/LoopConcepts/CompilerErrors.f90
program LoopErrors
    implicit none
    
    integer :: i, j, k, counter, value
    integer :: arr(5)
    
    ! ERROR: Missing 'do' keyword
    i = 1, 10
        write(*,*) i
    end do

    ! ERROR: Using 'for' instead of 'do' (from C-style languages)
    for (i = 1, 10)
        write(*,*) i
    end do

    ! ERROR: Using parentheses around loop range (C-style)
    do (i = 1, 10)
        write(*,*) i
    end do

    ! ERROR: Missing 'end do'
    do i = 1, 10
        write(*,*) i

    ! ERROR: Using 'end loop' instead of 'end do'
    do i = 1, 10
        write(*,*) i
    end loop

    ! ERROR: Using semicolons instead of commas in range
    do i = 1; 10
        write(*,*) i
    end do

    ! ERROR: Using 'to' instead of comma (from other languages)
    do i = 1 to 10
        write(*,*) i
    end do

    ! ERROR: Using '..' for range (from Ada/others)
    do i = 1 .. 10
        write(*,*) i
    end do

    ! ERROR: Missing comma in range specification
    do i = 1 10
        write(*,*) i
    end do

    ! ERROR: Using assignment operator = instead of comma
    do i = 1 = 10
        write(*,*) i
    end do

    ! ERROR: Using ++ or += for increment (C-style)
    do i = 1, 10
        write(*,*) i
        i++
    end do

    ! ERROR: Missing 'while' keyword in do while loop
    counter = 0
    do (counter < 10)
        write(*,*) counter
        counter = counter + 1
    end do

    ! ERROR: Using 'loop' instead of 'do' (Ada style)
    loop i = 1, 10
        write(*,*) i
    end loop

    ! ERROR: Using 'break' instead of 'exit'
    do i = 1, 10
        if (i == 5) break
        write(*,*) i
    end do

    ! ERROR: Using 'continue' for loop control (wrong context)
    do i = 1, 10
        if (i == 5) continue
        write(*,*) i
    end do

    ! ERROR: Missing 'end if' inside loop
    do i = 1, 10
        if (i > 5) then
            write(*,*) 'Greater than 5'
        write(*,*) i
    end do

    ! ERROR: Using == instead of .eq. or proper comparison
    do i = 1, 10
        if (i == 5) then
            exit
        end if
    end do

    ! ERROR: Missing 'then' keyword in if statement
    do i = 1, 10
        if (i > 5)
            write(*,*) 'Greater'
        end if
    end do

    ! ERROR: Using curly braces instead of do/end do
    do i = 1, 10 {
        write(*,*) i
    }

    ! ERROR: Missing space before step value
    do i = 1, 10,2
        write(*,*) i
    end do

    ! ERROR: Using 'step' keyword (VB/other languages)
    do i = 1, 10, step 2
        write(*,*) i
    end do

    ! ERROR: Using 'by' for increment (not valid Fortran)
    do i = 1, 10 by 2
        write(*,*) i
    end do

    ! ERROR: Wrong order for reverse loop (missing negative step)
    do i = 10, 1
        write(*,*) i
    end do

    ! ERROR: Using 'downto' (Pascal style)
    do i = 10 downto 1
        write(*,*) i
    end do

    ! ERROR: Using 'in' keyword (Python/Ruby style)
    arr = [1, 2, 3, 4, 5]
    do i in arr
        write(*,*) i
    end do

    ! ERROR: Using 'foreach' (C#/other languages)
    foreach i in arr
        write(*,*) i
    end foreach

    ! ERROR: Missing opening 'do' statement with just condition
    while (counter < 10)
        write(*,*) counter
        counter = counter + 1
    end do

    ! ERROR: Using 'wend' instead of 'end do' (BASIC style)
    do while (counter < 10)
        write(*,*) counter
        counter = counter + 1
    wend

    ! ERROR: Using := for assignment
    do i := 1, 10
        write(*,*) i
    end do

    ! ERROR: Using floating point in loop range
    do i = 1.0, 10.0
        write(*,*) i
    end do

    ! ERROR: Using real variable as loop counter
    real :: x
    do x = 1.0, 10.0
        write(*,*) x
    end do

    ! ERROR: Missing condition in do while
    do while
        write(*,*) 'Hello'
        if (counter > 10) exit
    end do

    ! ERROR: Using parentheses incorrectly in range
    do i = (1), (10)
        write(*,*) i
    end do

    ! ERROR: Using 'until' instead of 'while' (other languages)
    do until (counter >= 10)
        write(*,*) counter
        counter = counter + 1
    end do

    ! ERROR: Missing parentheses around while condition
    do while counter < 10
        write(*,*) counter
        counter = counter + 1
    end do

    ! ERROR: Using 'loop' with 'end loop' (Ada style)
    loop
        write(*,*) 'Hello'
        if (counter > 10) exit
    end loop

    ! ERROR: Using 'repeat' (other languages)
    repeat
        write(*,*) counter
        counter = counter + 1
    until (counter >= 10)

    ! ERROR: Using label incorrectly with exit
    outer: do i = 1, 10
        inner: do j = 1, 10
            if (j == 5) exit outer
        end do inner
    end do outer

    ! ERROR: Wrong syntax for named loop construct
    outerloop do i = 1, 10
        write(*,*) i
    end do outerloop

    ! ERROR: Using ampersand incorrectly in loop
    do i = 1, 10 &
    write(*,*) i
    end do

    ! ERROR: Missing logical operators in compound condition
    do while (i < 10 j > 0)
        write(*,*) i
        i = i + 1
    end do

    ! ERROR: Using && or || instead of .and. .or.
    do while (i < 10 && j > 0)
        write(*,*) i
        i = i + 1
    end do

    ! ERROR: Using 'else if' as two words instead of 'else if'
    do i = 1, 10
        if (i < 5) then
            write(*,*) 'Less'
        elseif (i > 5) then
            write(*,*) 'Greater'
        end if
    end do

    ! ERROR: Missing colon after loop label
    myloop do i = 1, 10
        write(*,*) i
    end do myloop

    ! ERROR: Using C-style increment in step
    do i = 1, 10, i++
        write(*,*) i
    end do

    ! ERROR: Using array as loop bound without size()
    do i = 1, arr
        write(*,*) i
    end do

    ! ERROR: Using 'next' instead of 'end do' (BASIC/VB)
    do i = 1, 10
        write(*,*) i
    next i

    ! ERROR: Using 'end for' instead of 'end do'
    do i = 1, 10
        write(*,*) i
    end for

    ! ERROR: Missing increment specification entirely
    do i = 1, 10,
        write(*,*) i
    end do

end program LoopErrors
